crates/fj-operations/src/sketch.rs

```diff
@@ -56,18 +56,57 @@ impl Shape for fj::Sketch {
                 }
             }
             fj::Chain::PolyChain(poly_chain) => {
-                let points = poly_chain
-                    .to_segments()
-                    .into_iter()
-                    .map(|fj::SketchSegment::LineTo { point }| point)
-                    .map(Point::from);
-
-                let mut face = PartialFace::default();
-                face.exterior.write().surface = Partial::from(surface);
-                face.update_exterior_as_polygon_from_points(points);
-                face.color = Some(Color(self.color()));
-
-                face
+                let segments = poly_chain.to_segments();
+                assert!(
+                    !segments.is_empty(),
+                    "Attempted to compute a Brep from an empty sketch"
+                );
+
+                let exterior = {
+                    let mut cycle = PartialCycle {
+                        surface: Partial::from(surface),
+                        ..Default::default()
+                    };
+                    let mut line_segments = vec![];
+                    let mut arcs = vec![];
+                    poly_chain.to_segments().into_iter().for_each(
+                        |fj::SketchSegment { endpoint, route }| {
+                            let endpoint = Point::from(endpoint);
+                            match route {
+                                fj::SketchSegmentRoute::Direct => {
+                                    line_segments.push(
+                                        cycle
+                                            .add_half_edge_from_point_to_start(
+                                                endpoint,
+                                            ),
+                                    );
+                                }
+                                fj::SketchSegmentRoute::Arc { angle } => {
+                                    arcs.push((
+                                        cycle
+                                            .add_half_edge_from_point_to_start(
+                                                endpoint,
+                                            ),
+                                        angle,
+                                    ));
+                                }
+                            }
+                        },
+                    );
+                    line_segments.into_iter().for_each(|mut half_edge| {
+                        half_edge.write().update_as_line_segment()
+                    });
+                    arcs.into_iter().for_each(|(mut half_edge, angle)| {
+                        half_edge.write().update_as_arc(angle.rad())
+                    });
```

Comment:
I struggled with this one for a bit actually. It can't go in the same match statement because both ends of each `HalfEdge` need to be part of the polychain before these methods will work. `update_as_line_segment` updates the direction, and `update_as_arc` calculates the center point, radius, and angles; all of these are wrong unless both endpoints (i.e. the one from _this_ edge and also the _next_ edge) are already defined.

There's probably a better way to structure this. Maybe just storing the segment kind inside of the `HalfEdge` as another partial type, and then calculating all of the other metrics later? But I think that's definitely a followup.

Reply:
Ah, got it. Looks like I was a bit naive in my initial judgement :smile: 


> There's probably a better way to structure this. Maybe just storing the segment kind inside of the `HalfEdge` as another partial type, and then calculating all of the other metrics later? But I think that's definitely a followup.

Let's just leave it as-is for now. The structure of the various object and partial object types, as well as their relationships to each other are somewhat in flux anyway. Maybe a more elegant solution will present itself, but until then, this will do nicely.